src/zarr/config.py

```diff
@@ -2,16 +2,66 @@
 
 from typing import Any, Literal, cast
 
-from donfig import Config
+from donfig import Config as DConfig
 
+
+class BadConfigError(ValueError):
+    _msg = "bad Config: %r"
+
+
+class Config(DConfig):  # type: ignore[misc]
+    """Will collect configuration from config files and environment variables
+
+    Example environment variables:
+    Grabs environment variables of the form "ZARR_FOO__BAR_BAZ=123" and
+    turns these into config variables of the form ``{"foo": {"bar-baz": 123}}``
+    It transforms the key and value in the following way:
+
+    -  Lower-cases the key text
+    -  Treats ``__`` (double-underscore) as nested access
+    -  Calls ``ast.literal_eval`` on the value
+
+    """
+
+    def reset(self) -> None:
+        self.clear()
+        self.refresh()
+
+
+"""
+The config module is responsible for managing the configuration of zarr and  is based on the Donfig python library.
+For selecting custom implementations of codecs, pipelines, buffers and ndbuffers, first register the implementations 
+in the registry and then select them in the config.
+e.g. an implementation of the bytes codec in a class "NewBytesCodec", requires the value of codecs.bytes.name to be 
+"NewBytesCodec".
+Donfig can be configured programmatically, by environment variables, or from YAML files in standard locations
+e.g. export ZARR_CODECS__BYTES__NAME="NewBytesCodec"
+(for more information see github.com/pytroll/donfig)
+Default values below point to the standard implementations of zarr-python
+"""
 config = Config(
     "zarr",
     defaults=[
         {
             "array": {"order": "C"},
             "async": {"concurrency": None, "timeout": None},
-            "codec_pipeline": {"batch_size": 1},
             "json_indent": 2,
+            "codec_pipeline": {
+                "path": "zarr.codecs.pipeline.BatchedCodecPipeline",
```

Comment:
Seeing this, I wonder if we should use an object with a `path` (or `class`?) for codecs as well. While we don't need it right now, this could be useful in the future for backwards-compat. We could also add that in the future, though. 
cc @jhamman @d-v-b

Reply:
Sorry for the back and forth on this!

Reply:
What I am proposing is 
```python
        {
            "array": {"order": "C"},
            "async": {"concurrency": None, "timeout": None},
            "json_indent": 2,
            "codec_pipeline": {
                "path": "zarr.codecs.pipeline.BatchedCodecPipeline",
                "batch_size": 1,
            },
            "codecs": {
                "blosc": { "path": "zarr.codecs.blosc.BloscCodec" },
                "gzip": { "path": "zarr.codecs.gzip.GzipCodec" },
                "zstd": { "path": "zarr.codecs.zstd.ZstdCodec" },
                "bytes": { "path": "zarr.codecs.bytes.BytesCodec" },
                "endian": { "path": "zarr.codecs.bytes.BytesCodec" },  # compatibility with earlier versions of ZEP1
                "crc32c": { "path": "zarr.codecs.crc32c_.Crc32cCodec" },
                "sharding_indexed": { "path": "zarr.codecs.sharding.ShardingCodec" },
                "transpose": { "path": "zarr.codecs.transpose.TransposeCodec" },
            },
            "buffer": "zarr.buffer.Buffer",
            "ndbuffer": "zarr.buffer.NDBuffer",
        }

```